http: Join worker threads before deleting work queue

laanwj opened the pull request:
This prevents a potential race condition if control flow ends up in
`ShutdownHTTPServer` before the thread gets to `queue->Run()`,
deleting the work queue while workers are still going to use it.

Meant to fix #12362.

laanwj committed:
http: Join worker threads before deleting work queue

This prevents a potential race condition if control flow ends up in
`ShutdownHTTPServer` before the thread gets to `queue->Run()`,
deleting the work queue while workers are still going to use it.

Meant to fix #12362.

Signed-off-by: Wladimir J. van der Laan <laanwj@gmail.com>

ryanofsky commented on src/httpserver.cpp:
```diff
@@ -449,6 +449,7 @@ bool UpdateHTTPServerLogging(bool enable) {
 
 std::thread threadHTTP;
 std::future<bool> threadResult;
+static std::vector<std::thread> threadHTTPWorkers;
```
Could follow g_thread_http_workers naming convention.

laanwj replied:
Can we please first make sure that this fixes the issue before jumping on naming issues

ryanofsky reviewed (commented):
utACK 2fefa3077e69dfdbccbd010cda5627787ec92bb8. Untested, but change makes sense. Detached threads are always a little suspicious.

TheBlueMatt commented on src/httpserver.cpp:
```diff
@@ -460,8 +461,7 @@ bool StartHTTPServer()
     for (int i = 0; i < rpcThreads; i++) {
-        std::thread rpc_worker(HTTPWorkQueueRun, workQueue);
-        rpc_worker.detach();
+        threadHTTPWorkers.emplace_back(std::thread(HTTPWorkQueueRun, workQueue));
```
No need to create the temporary just to move it into a new std::thread constructor.

laanwj replied:
Right

TheBlueMatt reviewed (commented):
Tested ACK 2fefa3077e69dfdbccbd010cda5627787ec92bb8 - my previous sleeps and valgrind are no longer able to find any issue here.

Would love to see @jnewbery confirm he cannot reproduce anymore as well.

laanwj commented on src/httpserver.cpp:
```diff
@@ -487,6 +487,10 @@ void StopHTTPServer()
     if (workQueue) {
         LogPrint(BCLog::HTTP, "Waiting for HTTP worker threads to exit\n");
         workQueue->WaitExit();
```
I think we don't need WaitExit() anymore with this?

TheBlueMatt replied:
It would seem not, indeed. Testing shows no difference with it removed.

laanwj replied:
Ok, will remove it in a separate commit.

laanwj committed:
http: Remove WaitExit from WorkQueue

This function, which waits for all threads to exit, is no longer needed
now that threads are joined instead.

Signed-off-by: Wladimir J. van der Laan <laanwj@gmail.com>

laanwj commented:
ok, renamed the variable and added a commit to remove WaitExit()

jnewbery commented:
I can confirm that I'm not able to reproduce the error when running with commit f94665466ed50e868c98b1a1c708ad5767727bb6.

ryanofsky commented on src/httpserver.cpp:
```diff
@@ -141,13 +140,6 @@ class WorkQueue
-    void WaitExit()
-    {
-        std::unique_lock<std::mutex> lock(cs);
-        while (numThreads > 0)
```
In commit "http: Remove WaitExit from WorkQueue"

It looks like the numThreads variable is unused now, so it might be worth removing as well.

laanwj replied:
Thanks, I removed those in a new commit.

ryanofsky reviewed (commented):
utACK f94665466ed50e868c98b1a1c708ad5767727bb6. Only changes were variable rename and temp std::thread / WaitExit remove suggestions above.

promag commented on src/httpserver.cpp:
```diff
@@ -486,7 +478,10 @@ void StopHTTPServer()
         LogPrint(BCLog::HTTP, "Waiting for HTTP worker threads to exit\n");
-        workQueue->WaitExit();
+        for (auto& thread: g_thread_http_workers) {
+            thread.join();
```
From the [documentation](http://www.cplusplus.com/reference/thread/thread/~thread/), I don't think this is needed, `clear()` below is enough.

TheBlueMatt replied:
No, if you delete a thread without either calling detach() or join(), terminate() is called. join() is not automatically called for you.

promag reviewed (commented):
utACK.

sipa commented:
utACK f94665466ed50e868c98b1a1c708ad5767727bb6. This is easier to read as well.

dcousens reviewed (approved):
utACK

laanwj committed:
http: Remove numThreads and ThreadCounter

The HTTP worker thread counter, as well as the RAII object that was used
to maintain it, is unused now, so can be removed.

Signed-off-by: Wladimir J. van der Laan <laanwj@gmail.com>

promag commented on src/httpserver.cpp:
```diff
@@ -73,34 +73,13 @@ class WorkQueue
     explicit WorkQueue(size_t _maxDepth) : running(true),
-                                 maxDepth(_maxDepth),
-                                 numThreads(0)
+                                 maxDepth(_maxDepth)
```
Nit, dunno about correct indentation here..

promag commented on src/httpserver.cpp:
```diff
@@ -486,7 +457,10 @@ void StopHTTPServer()
     if (workQueue) {
         LogPrint(BCLog::HTTP, "Waiting for HTTP worker threads to exit\n");
-        workQueue->WaitExit();
+        for (auto& thread: g_thread_http_workers) {
```
Nit, space before `:`.

promag commented on src/httpserver.cpp:
```diff
@@ -449,6 +420,7 @@ bool UpdateHTTPServerLogging(bool enable) {
 
 std::thread threadHTTP;
 std::future<bool> threadResult;
```
Nit, this and the above could be static too.

laanwj replied:
Yes, but out of scope of this PR to change that.

promag reviewed (commented):
utACK 11e0151.

MeshCollider commented:
Nice simplification+fix!
utACK https://github.com/bitcoin/bitcoin/pull/12366/commits/11e01515fe0fbc7823d4111ad6e016a02c485a78

laanwj commented:
> Nice cleanup!

FWIW I wrote HTTPWorkQueue with the idea that it could likely be replaced with a 1-liner after c++11, I think there's scope for more code removal. Out of scope for a 0.16.0 fix ofc.
